### DISCUSSION TITLE
Should we model an existing modelling language?

### DISCUSSION THREAD

**Discussion by jondavid-black** [DevOps-MBSE/AaC#43] [Ideas]
AaC currently has a defined set of types based mostly on my experience in system architecture and uses language (i.e. keywords) that are intended to be approachable for new system modelers...easy to pick up and use quickly.

Maybe a better strategy would be to model something like SysML.  Chances are our target audience has already been trained in SysML at some point in the education or career.  Using familiar terminology could be a benefit to this audience and ease adoption.

[SysML Quick Reference](https://docs.nomagic.com/display/NMDOC/Quick+Reference+Guides?preview=/70389199/70389218/no-magic-quick-reference-guide_sysml.pdf)

An additional benefit would be the ability to clam AaC is "standards based"...which checks a lot of boxes for people who have or are trying to spend money on MBSE.

So here's the real question:  what experiment could we do to explore SysML modelling as code?

**Comment by Coffee2Bits** [comment]
What would be the target outcome of modeling SysML for AaC? Would it ensure that SysML represents a subset of AaC, or is the focus to make the AaC DSL vocabulary SysML-like?

As far as an experiment goes, if we're not replacing the AaC DSL with a SysML DSL then we could try leveraging plugins for the parser and validator that would allow AaC users to code their architecture in a SysML-oriented DSL that could then be used with AaC without forcing all AaC users to use a SysML-oriented DSL.

As far as an experimental model for the SysML as Code, I think trying to model the flow model in the SysML as code DSL would be a good challenge since flow is the most complex example system we have so far.

**Comment by jondavid-black** [comment]
The target outcome would be to make the AaC vocabulary SysML-like so that engineers already familiar with SysML could more easily adopt AaC for their work.

As an interesting side note, I see that SysML primitives are very limited.  The discussion about generating Protobuf may only get more challenging with SysML vocabulary. 

> SysML defines four primitive value types: String, Boolean, Integer, and Real. You can, of course, define your own primitive value types as specializations (subtypes) of these four.

**Comment by Coffee2Bits** [comment]
I don't think it'll be much of an impact on the protobuf work since the current AaC primitive types also require some further definition to map to protobuf types, just like the SysML primitives would.

The biggest impact will probably be any structural changes and any changes to the names of fields--since we rely heavily on maps and strings to access the parsed models.

I think a lower-impact experiment would still be to plugin-ize the parser and test out a SysML parser plugin. That being said, we'd still need to think about the parsed_model structure being different between AaC and SysML DSLs. So, if we want to support multiple data structures for the models then we may want to consider creating a DSL independent data structure to represent parsed models internally in AaC regardless of their DSL structure.

**Comment by jondavid-black** [comment]
Good point about the DSL challenges.  Maybe the best way to understand that would be to create a SysML.yaml and see what impact it creates on parsing and validation.  Is SysML even structured to allow for self-validation?

I think we should wait a bit to take any real action.  We don't want to disrupt our first sprint demo.  That will also give a little time to receive feedback from stakeholder interviews to factor into our thought processes.

**Comment by lizzcondrey** [comment]
Some work has been done around this discussion. I don't at all think this is ready to be closed, but worth checking in about where we are with this. Below are related items that I've found. 
#477 
#570
#549 
#310 
#551 
#491

**Comment by jondavid-black** [comment]
I know we had talked about this a good bit in the past, and I thought we had some capability for SysML in place.  But when I look at the repo I don't see it.  Is that work in a separate branch somewhere?

**Reply by lizzcondrey** [reply]
Correct, @Coffee2Bits has an integration branch for SysML located [here](https://github.com/DevOps-MBSE/AaC/tree/sysml-expiremental)
